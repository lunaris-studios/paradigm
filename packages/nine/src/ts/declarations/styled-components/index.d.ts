import * as Protocol from "@paradigmjs/protocol";

import "styled-components";
import "styled-components/cssprop";

declare module "styled-components" {
	/**
	 * Intended declaration merge for extension of the root theme
	 * in all `styled-components`.
	 *
	 * https://styled-components.com/docs/api#create-a-declarations-file
	 */
	export interface DefaultTheme extends Protocol.Theme {}
}
