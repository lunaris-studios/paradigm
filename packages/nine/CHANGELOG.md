# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 0.0.11 (2020-08-01)


### Bug Fixes

* **gatsby:** replaced * declaration glob capture to explication name reference ([a3d26a4](https://github.com/lunaris-studios/paradigm/commit/a3d26a4aab9c0eaa4763d92271dbfd99e210604f))





## 0.0.10 (2020-08-01)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.9 (2020-08-01)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.8 (2020-08-01)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.7 (2020-07-31)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.6 (2020-07-29)


### Bug Fixes

* **universal:** added 3-9 paramter overloads for compose ([5156c77](https://github.com/lunaris-studios/paradigm/commit/5156c770b7ea98a668d811e4e7c728eb0c6aef96))





## 0.0.5 (2020-07-29)


### Bug Fixes

* **sensors:** renamed exported locale sensor HOC withLocale -> withLocaleSensor ([6b2f186](https://github.com/lunaris-studios/paradigm/commit/6b2f186b6bed6938204ef8ea1c54592b486ff141))





## 0.0.4 (2020-07-29)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.3 (2020-07-28)


### Bug Fixes

* testing minor version bump ([f6a06ce](https://github.com/lunaris-studios/paradigm/commit/f6a06cead3ddfc86cd3465e15646f0b667294d27))





## 0.0.2 (2020-07-28)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.1 (2020-07-28)

**Note:** Version bump only for package @paradigmjs/nine





## 0.0.2 (2020-07-28)

**Note:** Version bump only for package @paradigmjs/nine
