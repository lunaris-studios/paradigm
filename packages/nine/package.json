{
	"name": "@paradigmjs/nine",
	"version": "0.0.15",
	"description": "React library to create 9-slice image surfaces",
	"repository": {
		"type": "git",
		"url": "git@github.com:lunaris-studios/paradigm.git",
		"directory": "packages/nine"
	},
	"license": "MIT",
	"author": "Sam Craig <sam@lunaris.io>",
	"main": "dist/index.js",
	"module": "dist/index.es.js",
	"types": "dist/index.d.ts",
	"files": [
		"dist"
	],
	"scripts": {
		"clean": "run-p \"clean:*\"",
		"clean:build": "rm -rf ./dist",
		"clean:debug": "rm -rf ./npm-debug*",
		"clean:modules": "rm -rf ./node_modules package-lock.json",
		"compile": "run-p \"compile:*\"",
		"compile:jest": "npm run test:jest-generate-output",
		"compile:rollup": "rollup -c",
		"dev": "run-p \"dev:*\"",
		"dev:jest": "npm run test:jest-generate-output -- --watch",
		"dev:rollup": "rollup -cw",
		"dist": "run-s \"dist:*\"",
		"dist:rollup": "rollup -c",
		"lint": "run-p \"lint:*\"",
		"lint-fix": "eslint --fix './src/**/*.{js,jsx,ts,tsx}'",
		"lint:es": "eslint './src/**/*.{js,jsx,ts,tsx}'",
		"start": "npm run dev",
		"test": "run-s test:type-check test:jest",
		"test:jest": "jest",
		"test:jest-generate-output": "jest --json --outputFile=./coverage/.jest-test-results.json || true",
		"test:type-check": "tsc",
		"verify": "npm-run-all compile -p dist test lint"
	},
	"dependencies": {
		"@paradigmjs/core": "^0.0.15",
		"@paradigmjs/protocol": "^0.0.15",
		"react": "^16.13.1",
		"react-spring": "^9.0.0-rc.3",
		"styled-components": "^5.1.1"
	},
	"devDependencies": {
		"@rollup/plugin-alias": "^3.1.1",
		"@testing-library/jest-dom": "^5.11.2",
		"@testing-library/react": "^10.4.8",
		"@types/jest": "^26.0.9",
		"@types/node": "^14.0.27",
		"@types/react": "^16.9.44",
		"@types/react-dom": "^16.9.8",
		"@types/testing-library__jest-dom": "^5.9.2",
		"@typescript-eslint/eslint-plugin": "^2.34.0",
		"@typescript-eslint/parser": "^2.34.0",
		"@zerollup/ts-transform-paths": "^1.7.18",
		"babel-eslint": "^10.1.0",
		"eslint": "^7.2.0",
		"eslint-config-airbnb": "^18.2.0",
		"eslint-config-airbnb-typescript": "^9.0.0",
		"eslint-config-prettier": "^6.11.0",
		"eslint-import-resolver-typescript": "^2.2.0",
		"eslint-plugin-eslint-comments": "^3.2.0",
		"eslint-plugin-flowtype": "^4.0.0",
		"eslint-plugin-import": "^2.22.0",
		"eslint-plugin-jest": "^23.20.0",
		"eslint-plugin-jsx-a11y": "^6.3.1",
		"eslint-plugin-prettier": "^3.1.4",
		"eslint-plugin-promise": "^4.2.1",
		"eslint-plugin-react": "^7.20.5",
		"eslint-plugin-react-hooks": "^4.0.0",
		"eslint-plugin-unicorn": "^21.0.0",
		"jest": "^26.2.2",
		"jest-extended": "^0.11.5",
		"npm-check-updates": "^7.0.3",
		"npm-run-all": "^4.1.5",
		"prettier": "^2.0.5",
		"rollup": "^2.23.0",
		"rollup-plugin-typescript2": "^0.27.2",
		"ts-jest": "^26.1.4",
		"ttypescript": "^1.5.10",
		"typescript": "^3.9.7"
	},
	"peerDependencies": {
		"react": "^16.13.1",
		"react-dom": "^16.13.1"
	},
	"publishConfig": {
		"access": "public"
	},
	"gitHead": "c4734a568a3c12e70a85faf416465e04a935a71e"
}
