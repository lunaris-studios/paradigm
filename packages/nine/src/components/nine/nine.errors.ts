import * as Neptune from "@paradigmjs/paradigm";

const ns = Neptune.DISPLAYNAME_PREFIX;

export const WARN_NO_HEIGHT = ns + ` <Nine> height prop will default to 100px`;
export const ERROR_NO_IMAGE =
	ns + ` <Nine> component failed to render, no image was provided for the 9-slice`;
export const WARN_NO_CORNER = ns + ` <Nine> cornerLength prop will default to 25px`;
export const WARN_NO_WIDTH = ns + ` <Nine> width prop will default to 100px`;
