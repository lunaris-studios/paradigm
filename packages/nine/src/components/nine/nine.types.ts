import * as CSS from "csstype";

/**
 * Default height & width of the 9-slice image.
 */
export interface INineImageDimensions {
	height: number;
	width: number;
}

/**
 * Dimensions of a 9-slice section.
 */
export interface INineSectionDimensions {
	height: number;
	width: number;
}

/**
 * Coordinates on the x & y axis to offset the provided image for each section.
 */
export interface INineSectionBackgroundImageOffset {
	x: CSS.Property.BackgroundPositionX;
	y: CSS.Property.BackgroundPositionY;
}

/**
 * Cardinal positions used to denote the layout of 9-slice components.
 */
export enum NineCoordinate {
	NORTH = "NORTH",
	SOUTH = "SOUTH",
	EAST = "EAST",
	WEST = "WEST",
	NORTH_WEST = "NORTH_WEST",
	NORTH_EAST = "NORTH_EAST",
	SOUTH_WEST = "SOUTH_WEST",
	SOUTH_EAST = "SOUTH_EAST",
	CENTER = "CENTER",
}
