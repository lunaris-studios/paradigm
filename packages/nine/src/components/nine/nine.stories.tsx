import * as React from "react";

import * as Util from "~/util";

import * as Component from "./nine";

const Template: Util.Storybook.React.Story<Component.INineProps> = (args) => (
	<Component.Nine {...args} />
);

export const Basic = Template.bind({});
Basic.args = {};

export default {
	title: "Nine",
	component: Component.Nine,
} as Util.Storybook.React.Meta;
