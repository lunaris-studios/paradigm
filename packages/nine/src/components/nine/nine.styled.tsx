import * as Protocol from "@paradigmjs/protocol";
import * as SC from "styled-components";
import * as Spring from "react-spring";
// re-import `styled-components` development mode DOM classnames.
import styled, { css } from "styled-components";

import * as Util from "~/util";

import * as Component from "./nine";
import * as Types from "./nine.types";

/**
 * Table of Contents
 *
 * [Nine]
 */

/**
 * [Nine]
 * - [Nine.Container(WRAPPER)]
 * - - [Nine.Row(WRAPPER)(MAP<3>)]
 * - - - [Nine.Section(ELEMENT)(MAP<3>)]
 */

interface Nine {
	Container: Util.SC.Styled<Spring.AnimatedComponent<"div">, INineContainerProps>;
	Row: Util.SC.Styled<"div", INineRowProps>;

	StubImage: Util.SC.Styled<"img", INineStubImageProps>;
	Section: Util.SC.Styled<"div", INineSectionProps>;
}

export const Nine = {} as Nine;

/**
 * [Nine.Container]
 */

interface INineContainerProps {
	corner: Component.INineProps["corner"];
	height: Component.INineProps["height"];
	width: Component.INineProps["width"];
}

Nine.Container = styled(Spring.animated.div)<INineContainerProps>`
	${(props: INineContainerProps) => {
		const { height, width, corner } = props;

		const heightPx = Protocol.Snippets.value(height, Protocol.Unit.PX);
		const widthPx = Protocol.Snippets.value(width, Protocol.Unit.PX);
		const cornerPX = Protocol.Snippets.value(corner, Protocol.Unit.PX);

		return css`
			height: ${heightPx};
			width: ${widthPx};
			grid-template-columns: ${cornerPX} 1fr ${cornerPX};
			grid-template-rows: ${cornerPX} 1fr ${cornerPX};
		`;
	}}
	/** */
	display: grid;
	grid-template-areas:
		"${Types.NineCoordinate.NORTH_WEST} ${Types.NineCoordinate.NORTH} ${
	Types.NineCoordinate.NORTH_EAST
}"
		"${Types.NineCoordinate.WEST} ${Types.NineCoordinate.CENTER} ${Types.NineCoordinate.EAST}"
		"${Types.NineCoordinate.SOUTH_WEST} ${Types.NineCoordinate.SOUTH} ${
	Types.NineCoordinate.SOUTH_EAST
}";
`;

/**
 * [Nine.Row]
 */

interface INineRowProps {}

Nine.Row = styled("div")<INineRowProps>``;

/**
 * [Nine.StubImage]
 */

interface INineStubImageProps {}

Nine.StubImage = styled("img")<INineStubImageProps>`
	display: none;
`;

/**
 * [Nine.Section]
 */

interface INineSectionProps {
	coordinate: Types.NineCoordinate;
	corner: Component.INineProps["corner"];
	height: Component.INineProps["height"];
	image: Component.INineProps["image"];
	imageSize: Component.INineState["imageSize"];
	width: Component.INineProps["width"];
}

Nine.Section = styled("div")<INineSectionProps>`
	/* ${Protocol.Snippets.debug()} */
	/** */
	${(props: INineSectionProps) => {
		const { coordinate, corner, height, image, imageSize, width } = props;

		const bgImage = Boolean(coordinate !== Types.NineCoordinate.CENTER)
			? `url(${image})`
			: "";

		const bgPosition = getSectionBgPosition(coordinate);

		const bgSize = getSectionBgSize(coordinate, corner, imageSize, height, width);
		const bgSizeWidth = Boolean(bgSize.width)
			? Protocol.Snippets.value(bgSize.width, Protocol.Unit.PX)
			: "auto";
		const bgSizeHeight = Boolean(bgSize.height)
			? Protocol.Snippets.value(bgSize.height, Protocol.Unit.PX)
			: "auto";

		const display = Boolean(coordinate === Types.NineCoordinate.CENTER)
			? "contents"
			: "block";

		return css`
			background-image: ${bgImage};
			background-position-x: ${bgPosition.x};
			background-position-y: ${bgPosition.y};
			background-size: ${bgSizeWidth} ${bgSizeHeight};
			display: ${display};
			grid-area: ${coordinate};
		`;
	}}
`;

function getSectionBgSize(
	coordinate: Types.NineCoordinate,
	corner: INineSectionProps["corner"],
	imageSize: INineSectionProps["imageSize"],
	height: INineSectionProps["height"],
	width: INineSectionProps["width"],
): Types.INineSectionBgSize {
	const size: Types.INineSectionBgSize = {
		height: 0,
		width: 0,
	};

	const widthScaleOffset = Math.ceil((corner / imageSize!.width) * width);
	const heightScaleOffset = Math.ceil((corner / imageSize!.height) * height);

	switch (coordinate) {
		/** Top & Bottom */
		case Types.NineCoordinate.NORTH:
		case Types.NineCoordinate.SOUTH:
			size.height = imageSize!.height;
			size.width = width + widthScaleOffset;
			break;

		/** Left & Right */
		case Types.NineCoordinate.EAST:
		case Types.NineCoordinate.WEST:
			size.height = height + heightScaleOffset;
			size.width = imageSize!.width;
			break;

		case Types.NineCoordinate.CENTER:
			size.height = height - corner * 2;
			size.width = width - corner * 2;
	}

	return size;
}

function getSectionBgPosition(
	coordinate: Types.NineCoordinate,
): Types.INineSectionBackgroundImageOffset {
	const backgroundPosition: Types.INineSectionBackgroundImageOffset = {
		x: "center",
		y: "center",
	};

	switch (coordinate) {
		case Types.NineCoordinate.NORTH:
			backgroundPosition.x = "center";
			backgroundPosition.y = "top";
			break;

		case Types.NineCoordinate.SOUTH:
			backgroundPosition.x = "center";
			backgroundPosition.y = "bottom";
			break;

		case Types.NineCoordinate.EAST:
			backgroundPosition.x = "right";
			backgroundPosition.y = "center";
			break;

		case Types.NineCoordinate.WEST:
			backgroundPosition.x = "left";
			backgroundPosition.y = "center";
			break;

		case Types.NineCoordinate.NORTH_WEST:
			backgroundPosition.x = "left";
			backgroundPosition.y = "top";
			break;

		case Types.NineCoordinate.NORTH_EAST:
			backgroundPosition.x = "right";
			backgroundPosition.y = "top";
			break;

		case Types.NineCoordinate.SOUTH_EAST:
			backgroundPosition.x = "right";
			backgroundPosition.y = "bottom";
			break;

		case Types.NineCoordinate.SOUTH_WEST:
			backgroundPosition.x = "left";
			backgroundPosition.y = "bottom";
			break;

		default:
			break;
	}

	return backgroundPosition;
}
