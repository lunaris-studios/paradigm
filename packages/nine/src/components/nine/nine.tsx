import * as Abstract from "@paradigmjs/abstract";
import * as React from "react";
import * as Spring from "react-spring";
import styled, { css } from "styled-components";

import * as Common from "~/common";

import * as Errors from "./nine.errors";
import * as Styled from "./nine.styled";
import * as Types from "./nine.types";

export interface INineProps {
	/**
	 * Squared length reserved for the non-scalable, static corner elements.
	 * of the 9-slice layout.
	 * @default 8;
	 */
	corner: number;

	/**
	 * Passed as a child to the `Types.NineCoordinate.CENTER` 9-slice section.
	 */
	children?: React.ReactNode;

	/**
	 * Height of the surface.
	 * @default 128
	 */
	height: number;

	/**
	 * File path of the image to use for the background of the 9-slice
	 * surface.
	 */
	image: string;

	/**
	 * Styling to apply to the child element.
	 */
	style?: object;

	/**
	 * HTML tag to use for element.
	 * @default "div"
	 */
	tagName: keyof JSX.IntrinsicElements;

	/**
	 * Width of the surface.
	 * @default 128
	 */
	width: number;
}

export interface INineDefaultProps extends Omit<INineProps, "image" | "children"> {}

export interface INineState {
	/**
	 * Height & width of the 9-slice image.
	 * @default undefined
	 */
	imageSize: Types.INineImageSize | undefined;

	/**
	 * Indicates whether or not the image has finished loading.
	 * @default false
	 */
	isLoaded: boolean;
}

const defaultProps = Object.freeze<INineDefaultProps>({
	corner: 8,
	height: 128,
	tagName: "div",
	width: 128,
});

const defaultState = Object.freeze<INineState>({
	imageSize: undefined,
	isLoaded: false,
});

export class Nine extends Abstract.AbstractPureComponent<INineProps, INineState> {
	public static readonly displayName = `${Common.DISPLAYNAME_PREFIX}.Nine`;

	public constructor(props: INineProps) {
		super(props);

		this.onImgLoad = this.onImgLoad.bind(this);
	}

	public static readonly defaultProps: INineDefaultProps = defaultProps;
	public state: INineState = defaultState;

	public static Styled = Styled;
	public static Errors = Errors;
	public static Types = Types;

	public render(): JSX.Element {
		const { children, corner, height, image, tagName, width } = this.props;
		const { imageSize, isLoaded } = this.state;

		const sectionProps = { corner, height, image, imageSize, width };

		return (
			<React.Fragment>
				<Styled.Nine.StubImage src={image} onLoad={this.onImgLoad} />
				<Spring.Transition items={isLoaded} from={{ opacity: 0 }} enter={{ opacity: 1 }}>
					{(springs, isLoaded) =>
						isLoaded && (
							<Styled.Nine.Container
								corner={corner}
								height={height}
								style={springs}
								width={width}
							>
								{/* Row One */}
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.NORTH_WEST}
								/>
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.NORTH}
								/>
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.NORTH_EAST}
								/>

								{/* Row Two */}
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.WEST}
								/>
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.CENTER}
								>
									{children}
								</Styled.Nine.Section>
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.EAST}
								/>

								{/* Row Three */}
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.SOUTH_WEST}
								/>
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.SOUTH}
								/>
								<Styled.Nine.Section
									{...sectionProps}
									coordinate={Types.NineCoordinate.SOUTH_EAST}
								/>
							</Styled.Nine.Container>
						)
					}
				</Spring.Transition>
			</React.Fragment>
		);
	}

	/** */

	protected validateProps(props: INineProps): void {
		const { height, image, corner, width } = props;

		if (height == null) {
			console.warn(Errors.WARN_NO_HEIGHT);
		}
		if (image == null) {
			console.error(Errors.ERROR_NO_IMAGE);
		}
		if (corner == null) {
			console.warn(Errors.WARN_NO_CORNER);
		}
		if (width == null) {
			console.warn(Errors.WARN_NO_WIDTH);
		}
	}

	public onImgLoad(event: React.SyntheticEvent<HTMLImageElement, Event>): void {
		const { currentTarget: img } = event;

		this.setState({
			imageSize: {
				height: img.naturalHeight,
				width: img.naturalWidth,
			},
			isLoaded: true,
		});
	}
}
