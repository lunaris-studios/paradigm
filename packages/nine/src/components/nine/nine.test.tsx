import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";
import * as Testing from "@testing-library/react";
import * as SCTU from "styled-components/test-utils";

import * as Component from "./nine";
import * as Styled from "./nine.styled";

describe("Nine", () => {
	it("renders correctly", () => {
		const node = Testing.render(<Component.Nine image="stub" />);
		const NineContainer = SCTU.find(node.baseElement, Styled.Nine.Container);

		expect(NineContainer).toMatchSnapshot();
	});
});
