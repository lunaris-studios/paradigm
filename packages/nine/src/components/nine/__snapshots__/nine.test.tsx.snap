// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nine renders correctly 1`] = `
<div
  class="sc-AxjAm iTxQMi"
  height="128"
  width="128"
>
  <div
    class="sc-AxhCb dASrgk"
  />
  <div
    class="sc-AxhCb eZhbGV"
  />
  <div
    class="sc-AxhCb jgvTtt"
  />
  <div
    class="sc-AxhCb lfLDmv"
  />
  <div
    class="sc-AxhCb Idjxq"
  />
  <div
    class="sc-AxhCb cijAeO"
  />
  <div
    class="sc-AxhCb iZYjum"
  />
  <div
    class="sc-AxhCb ftKOyX"
  />
  <div
    class="sc-AxhCb jdDrLT"
  />
</div>
`;
